### ISSUE TITLE
Pilot parameters changes

### ISSUE THREAD

**Issue by simonbyrne** [CliMA/Thermodynamics.jl#82]
Following discussion in https://github.com/CliMA/CLIMAParameters.jl/issues/59, it would be worth trying the following:

1. Create 3 thermodynamics structs to indicate the current regime 
- `DryThermodynamics`
- `MoistEquilThermodynamics`
- `MoistNonEquilThermodynamics`
These should have fields for the necessary parameters they need (e.g. `DryThermodynamics` won't require the moisture-related parameters)

2. Add types for the corresponding state for each regime: initially these can just be the `Phase` structs, with the parameter sets removed, e.g.
```
struct PhaseDry{FT}
    "internal energy"
    e_int::FT
    "density of dry air"
    ρ::FT
end
```
 - should we rename "phase" to "state" in these names?
 - we could even allow for different ways to specify state, or even get rid of these in favor of named tuples?

3. Pass add an extra argument indicating the regime, e.g. `air_temperature(ts::ThermodynamicState)` would become `air_temperature(th::Thermodynamics, ts::ThermodynamicState)`

(we can leave the profiles out for the time being)

4. Integrate (1) with the .toml file loading discussed in CLIMAParameters.jl.

**Comment by odunbar** [comment]
I'm on it! 

I think this may also help add readability to how we address different cases with thermodynamics too, The current setup relies on zero-ing out the math for dry cases, with a default argument 
```julia
q = PhasePartition(tot=0,liq=0,ice=0)
```
rather than having different methods for dry cases. An example is for `air_temperature(...)`, runs the following
```julia
function air_temperature(
    param_set::APS,
    e_int::FT,
    q::PhasePartition{FT} = q_pt_0(FT),
) where {FT <: Real}
    T_0::FT = CPP.T_0(param_set)
    e_int_v0::FT = CPP.e_int_v0(param_set)
    e_int_i0::FT = CPP.e_int_i0(param_set)
    return T_0 +
           (
        e_int - (q.tot - q.liq) * e_int_v0 + q.ice * (e_int_v0 + e_int_i0)
    ) / cv_m(param_set, q)
end
```
thus still calls `e_int_v0`, `e_int_i0`, which are related to vapor and ice, and requires definition of `q.tot` etc. (defaults being 0). 
If we want to use our new parameter setup we will require methods:
```julia
function air_temperature(dt::DryThermodynamics, ts::ThermodynamicState) ... end
function air_temperature(met::MoistEquilibriumThermodynamics, ts::ThermodynamicState) ... end
function air_temperature(mnt::MoistNonequlibriumThermodynamics, ts::ThermodynamicState) ... end
```
each depending only on relevant parameters given by the first argument.

**Comment by charleskawczynski** [comment]
I would recommend leaving the original form of `air_temperature(::APS,::FT,::PhasePartition{FT})` to start with.. The idea of splitting this out is nice but, IMO, it's PMO and will bloat the codebase (and tests, we'll need separate tests for all of these additional methods). The design of the latter, user-facing, methods looks fine. Might I suggest still using the name "parameters" in the struct name?

> should we rename "phase" to "state" in these names?

Yes! I support this!

**Comment by odunbar** [comment]
I agree this requires more methods, but the example of air temperature above does demonstrate a different form of bloat though in the current base. That is, we currently 
(i) need to define moisture-related parameters to run a dry model, and 
(ii) we currently need to define  functions (e.g. in the above, we also need to define `cv_m(...)`) for dry air, which again is a moisture related function so shouldn't need a dry air definition?

**Comment by tapios** [comment]
We will rarely need a dry model--we are building a climate model. And when we do, as @charleskawczynski says, defining dry as the special case `q=0` is just fine (and in my view preferable, since it ensures that the dry limit of the moist model, i.e., `q \to 0`) converges to the dry case).

**Comment by odunbar** [comment]
OK after some more conversations - I'm going to sort the parameters first in a PR. Refactoring can be done further later - but I want to get this going ASAP so we can progress with TurbulenceConvection/Microphysics thereafter.

The plan
1. Create a parameter file (in thermodynamics) and read from there
2. Create a **single** `ThermodynamicsParameters` struct for parameters. As on further inspection (and @tapios @charleskawczynski comments) we don't really dispatch often over the `Dry`, `MoistEquilibrium` cases, thus it doesn't really make sense to subdivide parameter sets.
3. Have all local parameters get functions from `ThermodynamicsParameters` following the naming convention in the google sheet.
(4. Remove parameters from the phase, this will also allow the possibility of renaming it to "state")

**Comment by charleskawczynski** [comment]
Closed by #100. We can open a new issue on further optimizing parameter box for specific types, but we should make profile-guided decisions on these optimizations.
